### PR TITLE
[Issue 24031]: block editor initialize remote extensions from field variables

### LINKED ISSUE #24031: Block Editor: Initialize remote extensions from field variables
### User Story

As a user, I want to be able to add custom extensions from remote sources into a block editor field.

### Acceptance Criteria

- [ ] Get remote extensions from field variables settings object
- [ ] Load the extensions from remote sources (for test you can upload the file to your dotCMS) into the block editor field.

### Proposed Objective

Core Features

### Proposed Priority

Priority 2 - Important

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

_No response_

### Assumptions & Initiation Needs

An example of the settings object

```JavaScript
{
  extensions: [
    {
      url: "https://....",
      actions: [
        {
          command: 'insertProduct', // either insert or functionality
          menuLabel: 'Insert Product',
          icon: 'product' // material design
        },
        {
          command: 'insertProduct', // either insert or functionality
          menuLabel: 'Insert Product',
          icon: 'product' // material design
        }, {
          command: 'insertProduct', // either insert or functionality
          menuLabel: 'Insert Product',
          icon: 'product' // material design
        },
        {
          command: 'insertProduct', // either insert or functionality
          menuLabel: 'Insert Product',
          icon: 'product' // material design
        }
      ]
    },
    {
      url: "https://....",
      actions: [
        {
          command: 'insertProduct', // either insert or functionality
          menuLabel: 'Insert Product',
          icon: 'product' // material design
        },
        {
          command: 'insertProduct', // either insert or functionality
          menuLabel: 'Insert Product',
          icon: 'product' // material design
        }, {
          command: 'insertProduct', // either insert or functionality
          menuLabel: 'Insert Product',
          icon: 'product' // material design
        },
        {
          command: 'insertProduct', // either insert or functionality
          menuLabel: 'Insert Product',
          icon: 'product' // material design
        }
      ]
    }
  ]

}
```

Actions will be taking care in this issue: https://github.com/dotCMS/core/issues/24032

### Sub-Tasks & Estimates

_No response_

### PR DESCRIPTION
**Pull request by manuelrojas**
### Proposed Changes
* change 1
* change 2

### Checklist
- [ ] Tests
- [ ] Translations
- [ ] Security Implications Contemplated (add notes if applicable)

### Additional Info
CustomBlocks variable to test:
```
{
	"extensions": [{
		"url": "http://localhost:8080/application/customBlocks.js",
		"actions": [{
				"name": "CustomExtension",
				"command": "insertProduct",
				"menuLabel": "Insert Product",
				"icon": "product"
			},
			{
				"name": "HighlightCustom",
				"command": "insertProduct",
				"menuLabel": "Insert Product",
				"icon": "product"
			},
			{
				"name": "CustomNode",
				"command": "insertProduct",
				"menuLabel": "Insert Product",
				"icon": "product"
			}
		]
	}]
}
```
### Screenshots
Original             |  Updated
:-------------------------:|:-------------------------:
** original screenshot **  |  image
  | image

### CODE REVIEW

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/block-editor/src/lib/components/dot-block-editor/dot-block-editor.component.html [hunk 1]
```diff
@@ -1,9 +1,9 @@
-<div class="editor-wrapper" [style]="customStyles">
-    <tiptap-editor [(ngModel)]="content" [editor]="editor" (keyup)="subject.next()"></tiptap-editor>
+<div class="editor-wrapper" *ngIf="editor" [style]="customStyles">
+    <tiptap-editor [(ngModel)]="content" [editor]="editor"></tiptap-editor>
 </div>
 
 <dot-editor-count-bar
-    *ngIf="showCharData"
+    *ngIf="showCharData && editor"
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/components/dot-block-editor/dot-block-editor.component.html:6]: Since we are using the conditional `*ngIf="editor"` in both elements, we wrap them with an `ng-container` to use the conditional only once:

```html
<ng-container *ngIf="editor">
   <div>....</div>
   <dot-editor-count-bar>...</dot-editor-count-bar>
</ng-container>

```

File: core-web/libs/block-editor/src/lib/components/dot-block-editor/dot-block-editor.component.ts [hunk 2]
```diff
@@ -40,6 +48,22 @@ function toTitleCase(str) {
...
+
+type Block = {
+    url: string;
+    actions: Array<Actions>;
+};
+
+type CustomBlock = {
+    extensions: Block[];
+};
+
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/components/dot-block-editor/dot-block-editor.component.ts:66]: We can move these types to this [file](https://github.com/dotCMS/core/blob/0dea6164a7ee2ab47a5c3bcce49a76aed8776812/core-web/libs/dotcms-models/src/lib/dot-block-editor.model.ts)

File: core-web/libs/block-editor/src/lib/components/dot-block-editor/dot-block-editor.component.ts [hunk 3]
```diff
@@ -68,6 +93,7 @@ export class DotBlockEditorComponent implements OnInit, OnDestroy {
 
         this.setEditorJSONContent(content);
     }
+    @Output() valueChange = new EventEmitter<string>();
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/components/dot-block-editor/dot-block-editor.component.ts:96]: Great!

File: core-web/libs/block-editor/src/lib/components/dot-block-editor/dot-block-editor.component.ts [hunk 4]
```diff
@@ -130,6 +169,37 @@ export class DotBlockEditorComponent implements OnInit, OnDestroy {
         this.editor.view.dispatch(tr);
     }
 
+    /**
+     * This methods get the customBlocks variable to retrieve the custom modules as Objects.
+     * Validates that there is customBlocks defined.
+     */
+
+    private async getCustomBlocks(): Promise<AnyExtension[]> {
+        let data = {};
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/components/dot-block-editor/dot-block-editor.component.ts:178]: I think we can do this. This way, we can avoid using `data as CustomBlock` in different places.

```suggestion
        let data: CustomBlock;
```

File: core-web/libs/block-editor/src/lib/components/dot-block-editor/dot-block-editor.component.ts [hunk 5]
```diff
@@ -130,6 +169,37 @@ export class DotBlockEditorComponent implements OnInit, OnDestroy {
...
+        (data as CustomBlock).extensions.forEach((extension) => {
+            blockNames.push(...extension.actions.map((item) => item.name));
+        });
+
+        const moduleObj = customModules.reduce(
+            (prevModule, module) => ({ ...prevModule, ...module }),
+            {}
+        );
+
+        return blockNames.map((block) => moduleObj[block]);
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/components/dot-block-editor/dot-block-editor.component.ts:200]: Great!

File: .gitignore [hunk 6]
```diff
@@ -103,6 +103,7 @@ node_modules/
 /core-web/karma_html
 /core-web/installs
 /core-web/target
+/core-web/~/dotcms/*
```
Comment (rjvelazco) [inline .gitignore:106]: Is this change needed?

**Review by rjvelazco — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 6 above)
Comment (manuelrojas) [reply .gitignore:106]: @rjvelazco Yes, Talking with @fmontes we want to keep this one in case it happens to anyone else.

Comment (rjvelazco) [reply .gitignore:106]: @manuelrojas  Sounds god!

File: core-web/libs/block-editor/src/lib/components/dot-block-editor/dot-block-editor.component.ts [hunk 7]
```diff
@@ -130,6 +155,37 @@ export class DotBlockEditorComponent implements OnInit, OnDestroy {
...
+            console.warn('JSON parse fails, please check the JSON format.');
+
+            return [];
+        }
+
+        const extensionUrls = data.extensions.map((extension) => extension.url);
+        const customModules = await this.loadCustomBlocks(extensionUrls);
+        const blockNames = [];
+
+        (data as CustomBlock).extensions.forEach((extension) => {
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/components/dot-block-editor/dot-block-editor.component.ts:177]:
```suggestion
        data.extensions.forEach((extension) => {
```

**Review by rjvelazco — APPROVED** [review]
